### ISSUE TITLE
Incorrect depth estimation when there is low LiDAR density

### ISSUE THREAD

**Issue by SebsBarbas** [wvangansbeke/Sparse-Depth-Completion#36]
Hi,
I was working with your neural network and I have faced some small issues when downsampling the LiDAR input to around 500 samples. In order to downsample the input, I used the function that you provide: `downsample`. Afterwards, I have tested how it predicts depth in the pixels where there is actual LiDAR information in the input and I have got the following results:

```shell
LiDAR depth: 2049.609375 groundtruth depth: 2049.609375 NN depth: 2097.265625
LiDAR depth: 1317.578125 groundtruth depth: 1303.125 NN depth: 729.6875
LiDAR depth: 2034.765625 groundtruth depth: 1994.921875 NN depth: 1766.015625
LiDAR depth: 1038.671875 groundtruth depth: 1038.671875 NN depth: 604.6875
LiDAR depth: 1788.28125 groundtruth depth: 1784.375 NN depth: 527.34375
LiDAR depth: 1769.921875 groundtruth depth: 1740.234375 NN depth: 1535.9375
LiDAR depth: 1423.828125 groundtruth depth: 1423.828125 NN depth: 986.71875
LiDAR depth: 1203.125 groundtruth depth: 1190.625 NN depth: 743.359375
LiDAR depth: 1762.890625 groundtruth depth: 1757.421875 NN depth: 1470.3125
LiDAR depth: 2573.4375 groundtruth depth: 2540.234375 NN depth: 2076.953125
LiDAR depth: 655.078125 groundtruth depth: 655.078125 NN depth: 539.453125
LiDAR depth: 1229.296875 groundtruth depth: 1229.296875 NN depth: 436.71875
```
I have also seen that for the normal LiDAR depth maps, the output is never the depth map's values but very similar ones in the same pixel position:

```shell
LiDAR depth: 1219.53125 groundtruth depth: 1219.53125 NN depth: 1220.3125
LiDAR depth: 2633.203125 groundtruth depth: 2633.203125 NN depth: 2554.296875
LiDAR depth: 1113.671875 groundtruth depth: 1113.671875 NN depth: 1115.625
LiDAR depth: 4438.28125 groundtruth depth: 4438.28125 NN depth: 4425.0
LiDAR depth: 627.734375 groundtruth depth: 618.75 NN depth: 632.03125
LiDAR depth: 1235.15625 groundtruth depth: 1233.984375 NN depth: 1235.546875
LiDAR depth: 1150.390625 groundtruth depth: 1150.390625 NN depth: 1148.4375
LiDAR depth: 2821.484375 groundtruth depth: 2821.484375 NN depth: 2819.140625
LiDAR depth: 1296.484375 groundtruth depth: 1296.484375 NN depth: 1284.765625
```

I believe it is an issue due to the lack of information density but I wanted to know if there is any way of improving these results with such a sparse depth information or if retraining with a sparser dataset (creating my own one) would improve the accuracy.
Thanks for your help.
Best,
Sebastián

**Comment by wvangansbeke** [comment]
Hi @SebsBarbas,

I see. Of course if makes sense that the results are worse when the LIDAR is downsampled. A good sanity check would be to see if the results are still better than simply training with RGB. This number should also be in the paper. Definitely try this.

**Comment by SebsBarbas** [comment]
Yes, but the issue is that if a pixel in the depth input has got information d(u, v) = x, that same pixel in the output of the network in the same pixel (u, v) will not have the same value x as in the input, but a very different one. I was expecting that the depth at a pixel (u, v) in the output would have the same value as in the pixel (u, v) in the input if this last one contained LiDAR information, but I have seen that this does not happen, specially the input depth has very little information (500 depth pixels).

These following points are a comparison between: the depth input at a pixel which contained information, the groundtruth depth in that same pixel and the output of the network in that same pixel for a depth input with 500 LiDAR points:

```shell
LiDAR depth: 2049.609375 groundtruth depth: 2049.609375 NN depth: 2097.265625
LiDAR depth: 1317.578125 groundtruth depth: 1303.125 NN depth: 729.6875
LiDAR depth: 2034.765625 groundtruth depth: 1994.921875 NN depth: 1766.015625
LiDAR depth: 1038.671875 groundtruth depth: 1038.671875 NN depth: 604.6875
LiDAR depth: 1788.28125 groundtruth depth: 1784.375 NN depth: 527.34375
LiDAR depth: 1769.921875 groundtruth depth: 1740.234375 NN depth: 1535.9375
LiDAR depth: 1423.828125 groundtruth depth: 1423.828125 NN depth: 986.71875
LiDAR depth: 1203.125 groundtruth depth: 1190.625 NN depth: 743.359375
LiDAR depth: 1762.890625 groundtruth depth: 1757.421875 NN depth: 1470.3125
LiDAR depth: 2573.4375 groundtruth depth: 2540.234375 NN depth: 2076.953125
LiDAR depth: 655.078125 groundtruth depth: 655.078125 NN depth: 539.453125
LiDAR depth: 1229.296875 groundtruth depth: 1229.296875 NN depth: 436.71875
```

My doubt is if this comes from the fact that the network has been trained with a denser depth image containing more information and if this problem can be solved with retraining the network with a sparser dataset. 
Thanks for your time,
Sebastián

**Comment by wvangansbeke** [comment]
Hi @SebsBarbas,

Yes indeed. When you change the sparsity, you have to retrain the whole network. The network has to learn how much it can rely on the noisy lidar. When the sparsity is very low, the network can probably learn a simple interpolation/averaging function. When the sparsity is very high, the network will rely more on the RGB image.

Hope this helps.

**Comment by SebsBarbas** [comment]
Hi @wvangansbeke ,
But independently on the density of Lidar, shouldn't the network always place a high trust in each LiDAR point (and define the depth of that point something very close to the Lidar value) since they are always of very good quality and of high confidence following the confidence map from the local network?

**Comment by wvangansbeke** [comment]
Hi @SebsBarbas,
I see. If you would only do late fusion, I think that makes sense. In the current code base however, also early fusion is performed. So I would expect you still need to retrain.

**Comment by SebsBarbas** [comment]
Hi @wvangansbeke ,
Thanks for your help and quick responses,
Sebastián
